src/Thentos/Frontend/Handlers.hs

```diff
@@ -110,20 +111,90 @@ userCreateConfirm = do
+    (userView, _) <- runForm "update" userUpdateForm
+    (emailView, _) <- runForm "update_email" emailUpdateForm
+    case result of
+        Nothing -> blaze $ userUpdatePage userView emailView passwordView
+        Just (oldPw, newPw) -> do
+            result' <- snapRunAction' clearance $ changePassword uid oldPw newPw
+            case result' of
+                Right () -> blaze "Password Changed!"
+                Left e -> logger INFO (show e) >> blaze (errorPage "user update failed.")
+
+emailUpdate :: FH ()
+emailUpdate = runWithUserClearance $ \ clearance uid -> do
+    (passwordView, _) <- runForm "update_password" passwordUpdateForm
+    (userView, _) <- runForm "update" userUpdateForm
+    (emailView, result) <- runForm "update_email" emailUpdateForm
+    case result of
+        Nothing -> blaze $ userUpdatePage userView emailView passwordView
+        Just newEmail -> do
+            result' <- snapRunAction' clearance $ requestUserEmailChange uid newEmail
+            case result' of
+                Right token -> do
+                    config :: ThentosConfig <- gets (^. cfg)
+                    feConfig <- gets (^. frontendCfg)
+                    liftIO $ sendEmailChangeConfirmMail (Tagged $ config >>. (Proxy :: Proxy '["smtp"]))
+                                newEmail
+                                (urlEmailChangeConfirm feConfig token)
+                    blaze emailSentPage
+                Left UserEmailAlreadyExists -> blaze emailSentPage
+                Left e -> logger INFO (show e) >> blaze (errorPage "email update failed.")
+  where
+    emailSentPage = "Please confirm your new email address through the email we just sent you"
+
+urlEmailChangeConfirm :: HttpConfig -> ConfirmationToken -> L.Text
+urlEmailChangeConfirm feConfig (ConfirmationToken token) =
+    cs (exposeUrl feConfig) <//> "/user/email_confirm?token="
+        <> (L.fromStrict . decodeUtf8 . urlEncode . encodeUtf8 $ token)
+
+
+emailUpdateConfirm :: FH ()
+emailUpdateConfirm = do
+    mToken <- (>>= urlDecode) <$> getParam "token"
+    let meToken = ConfirmationToken <$$> decodeUtf8' <$> mToken
+    case meToken of
+        Nothing -> blaze $ errorPage "Missing token"
+        Just (Left _unicodeError) -> blaze $ errorPage "Bad token"
+        Just (Right token) -> do
+            result <- snapRunAction' allowEverything $ confirmUserEmailChange token
+            case result of
+                Right () -> blaze $ "Email address changed"
+                Left NoSuchToken -> blaze $ errorPage "No such token"
+                Left e -> do
+                    logger WARNING (show e)
+                    blaze $ errorPage "Error when trying to change email address"
+
+runWithUserClearance :: (ThentosClearance -> UserId -> FH ()) -> FH ()
 runWithUserClearance = runWithUserClearance' ()
 
-runWithUserClearance' :: a -> (ThentosClearance -> FH a) -> FH a
+runWithUserClearance' :: a -> (ThentosClearance -> UserId -> FH a) -> FH a
```

Comment:
the fact that the function argument also wants a uid should probably be discussed in a comment.  (answer questions like: how is it filled?  why is this not done before the call to `runWithUserClearance`?)